### PR TITLE
TensorFlow autologging: change every_n_iter default to 1, always log last epoch

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

Currently, the default behavior for TensorFlow autologging is to only log metrics every 100 epochs. Many TensorFlow training sessions (especially tf.keras sessions) last far fewer than 100 epochs, resulting in valuable metrics being omitted. This behavior is also confusing to users who may not have read the API docs / defaults carefully.

Accordingly, this PR adjusts the default value for `every_n_iter` to `1`. It also ensures that metrics for the last training epoch are always logged, regardless of the epoch number. We are not concerned about excessive metric logging latency because: 1. the default behavior in TensorFlow for estimator models is to record TensorBoard summary metrics every 100 epochs already (i.e. this change is a no-op for the default metric logging behavior w.r.t TensorFlow estimators) 2. we've integrated batching capabilities for metrics logging (`BatchMetricsLogger`) to minimize the impact on training latency.

## How is this patch tested?

TF2 autologging unit / integration tests

## Release Notes

TensorFlow autologging: the default value of `every_n_iter` has been changed to `1`, and metrics will always be logged for the last training iteration.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [x] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/tensorflow_autolog/test_tensorflow2_autolog.py [hunk 1]
```diff
@@ -700,14 +723,6 @@ def test_tf_estimator_autolog_model_can_load_from_artifact(tf_estimator_random_d
     mlflow.tensorflow.load_model("runs:/" + tf_estimator_random_data_run.info.run_id + "/model")
 
 
-@pytest.mark.large
```
Comment (dbczumar) [inline tests/tensorflow_autolog/test_tensorflow2_autolog.py:703]: This test is no longer necessary since autologging config updates are implemented & tested via higher-level `autologging_utils` APIs. Removing it. This was failing due to line 708 after the changes to always log metrics for the last epoch were implemented.

File: mlflow/tensorflow.py [hunk 2]
```diff
@@ -619,12 +620,41 @@ def _flush_queue():
...
+            # For thread safety and to avoid modifying a list while iterating over it, we record a
+            # separate list of the items being flushed and remove each one from the metric queue,
+            # rather than clearing the metric queue or reassigning it (clearing / reassigning is
+            # dangerous because we don't block threads from adding to the queue while a flush is
+            # in progress)
+            snapshot = _metric_queue[:]
+            for item in snapshot:
+                _metric_queue.remove(item)
+
+            # NB: Most TensorFlow APIs use one-indexing for epochs, while tf.Keras
```
Comment (apurva-koti) [inline mlflow/tensorflow.py:633]: bruh...

File: mlflow/tensorflow.py [hunk 3]
```diff
@@ -619,12 +620,41 @@ def _flush_queue():
...
+            metrics_to_log_immediately = [
+                item for item in snapshot if (item[1].step - 1) % _LOG_EVERY_N_STEPS == 0
+            ]
+
+            metrics_by_run = _assoc_list_to_map(metrics_to_log_immediately)
+            for run_id, metrics in metrics_by_run.items():
+                try_mlflow_log(client.log_batch, run_id, metrics=metrics, params=[], tags=[])
+
+            # If metrics for the last training epoch were not logged due to filtering on
+            # `every_n_iter`, log them now
```
Comment (apurva-koti) [inline mlflow/tensorflow.py:646]: I'm a little confused, why exactly is this necessary? If they were not logged due to filtering on `every_n_iter`, isn't that the purpose of that filter?

Comment (dbczumar) [reply mlflow/tensorflow.py]: That's a fair point. I was concerned that users might miss metrics for the final training epoch, which is more likely to produce the best results compared to other randomly selected epochs. However, because we're defaulting to `every_n_iter=1`, it seems acceptable to drop metrics for the last epoch for cases where `every_n_iter != 1` since users are opting into this behavior & likely know what they're doing. I've removed the logic that was introduced to unconditionally log metrics for the last epoch, which simplifies things quite a bit. Thanks!

Comment (apurva-koti) [reply mlflow/tensorflow.py]: Sounds great!

File: mlflow/tensorflow.py [hunk 4]
```diff
@@ -619,12 +619,19 @@ def _flush_queue():
...
+            # rather than clearing the metric queue or reassigning it (clearing / reassigning is
+            # dangerous because we don't block threads from adding to the queue while a flush is
+            # in progress)
+            snapshot = _metric_queue[:]
+            for item in snapshot:
+                _metric_queue.remove(item)
+
+            metrics_by_run = _assoc_list_to_map(snapshot)
+            for run_id, metrics in metrics_by_run.items():
+                try_mlflow_log(client.log_batch, run_id, metrics=metrics, params=[], tags=[])
```
Comment (apurva-koti) [inline mlflow/tensorflow.py:634]: I am pretty sure this was the behavior before anyway, but this makes it a little more clear that if `try_mlflow_log` fails to log a metric x, that metric won't be in the queue anymore due to line 630 and so will never be logged.
Not blocking though since this was already the case.

**Review by apurva-koti — APPROVED** [review]
Thanks for the changes, looks a lot cleaner :)

File: mlflow/tensorflow.py (hunk 4 above)
Comment (dbczumar) [reply mlflow/tensorflow.py:634]: Yep! That's expected / is consistent with past behavior. Thanks for calling this out!
